icu4j/main/classes/core/src/com/ibm/icu/impl/units/ComplexUnitsConverter.java

```diff
@@ -0,0 +1,134 @@
+        unitConverters_ = new ArrayList<>();
+        for (int i = 0, n = units_.size(); i < n; i++) {
+            if (i == 0) { // first element
+                unitConverters_.add(new UnitConverter(inputUnit, units_.get(i), conversionRates));
+            } else {
+                unitConverters_.add(new UnitConverter(units_.get(i - 1), units_.get(i), conversionRates));
+            }
+        }
+    }
+
+    /**
+     * Returns true if the specified `quantity` of the `inputUnit`, expressed in terms of the biggest
+     * unit in the MeasureUnit `outputUnit`, is greater than or equal to `limit`.
+     * <p>
+     * For example, if the input unit is `meter` and the target unit is `foot+inch`. Therefore, this
+     * function will convert the `quantity` from `meter` to `foot`, then, it will compare the value in
+     * `foot` with the `limit`.
+     */
+    public boolean greaterThanOrEqual(BigDecimal quantity, BigDecimal limit) {
+        assert !units_.isEmpty() ;
+
+        // NOTE: First converter converts to the biggest quantity.
+        return unitConverters_.get(0).convert(quantity).multiply(BigDecimal.valueOf(1 + EPSILON)).compareTo(limit) >= 0;
+    }
+
+    /**
+     * Returns outputMeasures which is an array with the corresponding values.
+     * - E.g. converting meters to feet and inches.
+     * 1 meter --> 3 feet, 3.3701 inches
+     * NOTE:
+     * the smallest element is the only element that could have fractional values. And all
+     * other elements are floored to the nearest integer
+     */
+    public Pair<ArrayList<Measure>, ArrayList<Pair<MeasureUnitImpl, BigDecimal>>> convert(BigDecimal quantity) {
+        ArrayList<Measure> result = new ArrayList<>();
+        ArrayList<Pair<MeasureUnitImpl, BigDecimal>> tempResult = new ArrayList<>();
+
+        for (int i = 0, n = unitConverters_.size(); i < n; ++i) {
+            quantity = (unitConverters_.get(i)).convert(quantity);
+
+            if (i < n - 1) {
+                // The double type has 15 decimal digits of precision. For choosing
+                // whether to use the current unit or the next smaller unit, we
+                // therefore nudge up the number with which the thresholding
+                // decision is made. However after the thresholding, we use the
+                // original values to ensure unbiased accuracy (to the extent of
+                // double's capabilities).
+                Number newQuantity = Math.floor(quantity.doubleValue() * (1 + EPSILON));
+
+                // NOTE: Measure would own its MeasureUnit. // TODO?
+                //  result.add(new Measure(newQuantity, units_.get(i).build()));
+                tempResult.add(Pair.of(units_.get(i), BigDecimal.valueOf(newQuantity.doubleValue())));
+
+                // Keep the residual of the quantity.
+                //   For example: `3.6 feet`, keep only `0.6 feet`
+                quantity = quantity.subtract(BigDecimal.valueOf(newQuantity.longValue()));
```

Comment:
It's possible we will still need the "if ((quantity - roundedQuantity) / quantity < DBL_EPSILON) {" logic, depending on how accurate we make our arbitrary precision calculations.

Can you add ComplexUnitConverterTest.java? Unit tests in "void UnitsTest::testComplexUnitsConverter() {" exercises the precision logic relatively well. (There might be a "TEST_EPSILON" which we can tweak to see how it relates to BigDecimal precision.)

Reply:
is it possible to just compare if it is less than zero, then make it zero

Reply:
For me, I think lets leave this part when we are doing the glue code. I am not sure if we needed it in this pull request.

Reply:
In my opinion it shouldn't be much more work to add testComplexUnitsConverter() to this PR instead of the glue code PR. Redoing a squash should be less than a minute, losing Shane's approve isn't a problem since I can approve on his behalf.

Me reviewing the validity of this "on behalf of the unit test" means I have to do more work, on behalf of the unit test. Meaning only adding the unit test in the glue code PR instead of this PR because "it's more work" means you would rather give me more work than you.

I can remove my review from this PR and just review the Glue Code PR instead.

Reply:
The logic looks reasonable. The unit-sorting logic is missing though, I'll start a new thread for that.